src/librustc_typeck/check/method/probe.rs

```diff
@@ -908,8 +957,26 @@ impl<'a, 'gcx, 'tcx> ProbeContext<'a, 'gcx, 'tcx> {
     fn pick_core(&mut self) -> Option<PickResult<'tcx>> {
         let steps = self.steps.clone();
 
-        // find the first step that works
-        steps.iter().filter_map(|step| self.pick_step(step)).next()
+        match self.looking_for {
```

Comment:
Can you keep the comment here, and explain why these two arms are different? (I agree it makes sense to treat them differently, but it won't make sense to me later when I've forgotten what's going on =)

Something like:

```
// Normally, we stop at the first step where we find a positive match.
// But when we are scanning for methods with a suitable return type,
// these methods have distinct names and hence may not shadow one another
// (also, this is just for hints, so precision is less important).
```

Note though that if you find two methods with the same name, but one of them is "more steps" later, we might try to be smart about that, since it would be harder for the user to call it (they would need to use some manual derefs, and we might want to show them that).